syntax = "proto3";

option go_package = "gitlab.cs.washington.edu/assafv/fs3/protos/fs3";

package fs3;

service fs3 {
  rpc Copy(CopyRequest) returns (CopyReply) {}
  rpc Remove(RemoveRequest) returns (RemoveReply) {}
  rpc Get(GetRequest) returns (GetReply) {}
}

enum Status {
  GREAT_SUCCESS = 0;
  NOT_FOUND = 1;
  INTERNAL_ERROR = 2;
  ILLEGAL_PATH = 3;
}

/* requests */

/* request used to copy a file from client to server */

message CopyRequest {
  string file_path = 1;
  bytes file_content = 2;
}

/* request used to remove a file from the server */

message RemoveRequest {
  string file_path = 1;
}

/* request used to get a file from the server, contents sent back to client */

message GetRequest {
  string file_path = 1;
}

/* replies */

/* reply containing status of copying file to the server */

message CopyReply {
  Status status = 1;
  string file_path = 2;
}

/* reply containing status of removing file from the server */

message RemoveReply {
  Status status = 1;
  string file_path = 2;
}

/* reply containing file contents gotten from the app server
  and an error status indicating success or error */

message GetReply {
  Status status = 1;
  string file_path = 2;
  bytes file_content = 3;
}

